### ISSUE TITLE
KDE 5 applications are slow

### ISSUE THREAD

**Issue by ttuegel** [NixOS/nixpkgs#21345]
The KDE 5 tests on Hydra regularly time-out due to KDE 5 applications being so slow. This is mainly due to searching so many paths in `XDG_DATA_DIRS`, as @edolstra pointed out [here](https://github.com/NixOS/nixpkgs/commit/b2e80a53cab7890024036f373a78ab8a560b4285#commitcomment-20259832). I thought it would be more appropriate to open a ticket for this issue because it has come up in different contexts. Let me review the problem itself and possible solutions.

## Cause

This problem affects all KDE 5 applications, but the application in particular which triggered this issue was Dolphin, so I will address it specifically. Dolphin is slow because loading icons requires searching every directory in `XDG_DATA_DIRS` for icons. Dolphin renders many icons, so this issue hits harder than other applications. `kdeWrapper` adds *many* paths to this list so that Dolphin and all the libraries it depends on may find their runtime dependencies. The wrapper allows the dependencies to be resolved whether the application is installed in the user or the system profile and it allows the application to run cleanly from the Nix store. It is also necessary to allow installation of applications depending on different versions of KDE Frameworks or even Qt.

## Solution (-1): symlink farming

Until recently, `kdeWrapper` worked by creating a symlink farm for all the runtime dependencies and adding that path to `XDG_DATA_DIRS`. This is theoretically nice: a filesystem is a tree, so this basically amounts to creating a lookup tree for all the runtime dependencies. The problem is, every symlink takes up one filesystem block; at 4 KB each, that adds up. As @edolstra pointed out in #20433, this can add up to ~50 MB per application and up to ~400 MB for a typical KDE 5 installation. This is obviously unacceptable!

## Solution 0: KDE unwrapped

If we remove the wrappers and resort to `propagatedUserEnvPkgs`, we gain the speed of symlink farming without the overhead. However, it will no longer be possible to run KDE 5 applications from the Nix store. As @abbradar pointed out, this will resurrect the long-standing problem of Qt applications "poisoning" the system against applications using other Qt versions. As a result, it will probably be impossible to install KDE 5 applications using `nix-env`. It will also break KDE 5 applications running outside of KDE, except by an obscure workaround.

## Solution 1: dynamic symlink farming

The problem with symlink farming is the filesystem block occupied by each symlink. Placing the symlinks on a block-less filesystem like tmpfs removes this overhead! This would require creating the symlink farm dynamically. That is non-negligible overhead, but in principle it only needs to be done once per store path per boot, so it may be an acceptable compromise. Unfortunately, it would be a security nightmare: it would essentially be possible for a non-privileged user to modify the `XDG_DATA_DIRS` of any running application at any time!

Since the only remotely-acceptable solution on this list is "just let KDE be slow," I'm open to any suggestions!

**Comment by abbradar** [comment]
Just my humble opinion (again, I'm not even using KDE!) but although all those options seem bad, I'd go with -1 instead of _status quo_ -- disk space is much cheaper than our time that we spend waiting applications to load. 400 MB is bad but not _as_ bad, and (as I've already mentioned in another comment) times would be much, much worse for people with spinning drives.

**Comment by ttuegel** [comment]
> times would be much, much worse for people with spinning drives.

I have spinning drives; it's not noticeable because of the filesystem cache, but on the virtual machine, the extra cache pressure causes a cascading failure.

Ironically, the huge size increase doesn't affect the graphical installer because squashfs is block-less! Now, if only there was a way to get a bunch of squashfs images mounted into the Nix store...

**Comment by abbradar** [comment]
@ttuegel Hm, sounds good but relying on cache still seems flaky to me. How long are cold boot times?

On the other hand my worries may be entirely ungrounded indeed -- let's bring in some data! I'll try to measure time of cold-booting Dolphin on my host and in a VM (with SSD). I don't have KDE so it would be a more or less good measure. Test plan: use xdotool to wait for window and close it, measure overall process time.

**Comment by ttuegel** [comment]
@abbradar Before you run any tests, let me check something about the symlink farm overhead. I think I can reduce it significantly because I was doing something wrong before.

**Comment by abbradar** [comment]
@ttuegel It's too late, I was interested in how to measure this ~_^.

Here's my script:
```
#!/bin/sh

for i in `seq 1 10`; do
  echo 3 | sudo tee /proc/sys/vm/drop_caches >/dev/null
  time -f 'ELAPSED: %e' result/bin/dolphin >/dev/null &
  while true; do
    win=$(xdotool getactivewindow)
    if xdotool getwindowname $win | grep -q Dolphin; then
      break
    fi
    sleep 0.05
  done
  killall -9 dolphin
done
```

Results for my machine:
```
ELAPSED: 1.91
ELAPSED: 1.77
ELAPSED: 1.79
ELAPSED: 1.87
ELAPSED: 1.78
ELAPSED: 1.78
ELAPSED: 1.87
ELAPSED: 1.87
ELAPSED: 1.87
ELAPSED: 1.88
```

This is big but manageable if this is a cold boot. I'll postpone measurements in a VM until your experiment with symlink farm. Notice that this is still a very crude measurement (`sleep 0.1` is used to wait).

**Comment by ttuegel** [comment]
I have a patch for smaller symlink farms now; ~10MB instead of ~50MB. I get these timings before:
```
ELAPSED: 6.46
ELAPSED: 6.75
ELAPSED: 5.95
ELAPSED: 6.27
ELAPSED: 6.58
ELAPSED: 6.52
ELAPSED: 6.49
ELAPSED: 6.27
ELAPSED: 6.45
ELAPSED: 6.47
```
and these timings after
```
ELAPSED: 4.08
ELAPSED: 4.63
ELAPSED: 4.45
ELAPSED: 4.64
ELAPSED: 4.87
ELAPSED: 4.60
ELAPSED: 4.35
ELAPSED: 4.86
ELAPSED: 4.63
ELAPSED: 4.67
```
That is a significant speed improvement!

@edolstra Is a ~10MB per-application overhead acceptable for a ~30% improvement in startup time?

**Comment by FRidh** [comment]
`kdeWrapper` is now used for each application. Would it make sense to instead say one would like to have an environment with certain KDE applications, and then symlink everything that is needed for that environment (`kde5.withPackages`). If you will have only one application in an env, the size increase will remain the same. If you have multiple applications in an env, I would expect a smaller size increase because libraries are symlinked only once.

**Comment by ttuegel** [comment]
> How about considering an env with KDE5 applications, and symlink everything that is needed for that environment (kde5.withPackages). If you will have only one application in an env, the size increase will remain the same. If you have multiple applications in an env, I would expect a smaller size increase because libraries are symlinked only once.

@FRidh This is possible. It is essentially what we do already for the Plasma desktop. Plasma consists of a collection of applications; rather than wrapping them individually, we wrap them into a single desktop. We can make this wrapper more expansive, but that will break application installation by `nix-env`.

**Comment by abbradar** [comment]
What about a hybrid approach? Make individually wrapped applications visible and installable in `nix-env` but also add `kde5.withPackages :: (PackageSet -> [Package]) -> Package` and promote it to be used in NixOS configurations?

**Comment by abbradar** [comment]
Maybe do it like this:

* `kde5.foobar` -- application `foobar` which is defined as: `foobar = wrapPackage (callPackage ./foobar.nix { });`;
* `wrapPackage` adds passthrough attribute `unwrapped`;
* `withPackages` takes a list of wrapped packages and re-wraps their `unwrapped` versions in a one big environment.

This way everybody can use `kde5.foobar` and those that want more optimized approach can throw it with other packages into `withPackages` and it will do the right thing.

**Comment by ttuegel** [comment]
> withPackages takes a list of wrapped packages and re-wraps their unwrapped versions in a one big environment.

How should `withPackages` know which targets to wrap?

**Comment by abbradar** [comment]
@ttuegel I'm sorry but I'm not sure what do you mean by targets -- maybe outputs? If yes then IIUC for applications we want `bin` or `out`, similarly to what we already do in `environment.systemPackages`.

**Comment by ttuegel** [comment]
> I'm sorry but I'm not sure what do you mean by targets -- maybe outputs?

`kdeWrapper` doesn't wrap every executable in the `bin` output, but only specific ones. Some packages have utilities programs that aren't intended to be user-accessible; it doesn't make sense to wrap these. I don't see how to make `withPackages` aware of application-specific logic like this.

`kdeWrapper` also does application-specific cycle breaking. For example, `dolphin-plugins` provides plugins for the Dolphin file manager. Building the plugins requires headers and libraries provided by the Dolphin build. The unwrapped Dolphin build cannot depend on the plugins, but the Dolphin wrapper needs to include the plugins! We can do this because the Dolphin expression has application-specific wrapper logic. If we defer making the wrapper to `withPackages`, I don't know how to communicate that logic.

**Comment by abbradar** [comment]
Ah, I see. What about just exposing `targets` and `paths` as passthru arguments? Then we have `wrapPackages packages = kdeWrapper { targets = concatMap (p: p.targets) packages; paths = concatMap (p: p.paths) packages); }`, and all the custom logic should apply (it's already collected in `kdeWrapper` and activated by having specific targets/paths, correct?).

**Comment by abbradar** [comment]
Hm, and even better: we can get rid of `wrapPackages` completely! In pseudo-code:

```
kdeWrapPackages pkgs =
  let targets = concatMap (p: p.targets) pkgs;
      paths = concatMap (p: p.paths) pkgs;
  in stdenv.mkDerivation {
    ...
    passthru = { inherit targets paths; };
  };
```

This way you can re-wrap packages endlessly:

`kdeWrapper [(kdeWrapper [a]) (kdeWrapper [b])] = kdeWrapper [a b]`

**Comment by ttuegel** [comment]
> This way you can re-wrap packages endlessly:

Not quite; each additional wrapper adds another level of symlink indirection! But your point is well taken: because Nix is not typed, we can essentially overload `kdeWrapper :: [Package] -> Package` with `kdeWrapper :: Package -> Package`.

**Comment by FRidh** [comment]
@abbradar That's exactly my though, pass through the unwrapped derivation as well as the arguments for the wrapper. 

> Not quite; each additional wrapper adds another level of symlink indirection! 

Why? If I understand correctly what @abbradar wrote and how I see it as well, is that only for the eventual environment (and that consist of multiple packages or a single) you create the symbolic links. 

In `kdeWrapper [(kdeWrapper [a]) (kdeWrapper [b])] = kdeWrapper [a b]` the only derivations built would be the unwrapped, and the eventual env with symbolic links/wrappers, right?

**Comment by abbradar** [comment]
I think there is misunderstanding of my idea, or I don't fully get how `kdeWrapper` works now. Sorry and please bear with me!

Let's define a KDE package (in some pseudo-typed Nix):
```
type KDEPackage = { targets :: [String], paths :: [Package] }
```

That is, a package (for the purposes of `kdeWrapper`) is two sets: of paths (that is, Nix packages) that we need to link and of targets (that is, names of binaries inside those paths).

To get a wrapped Nix package, we'd define a function:
```
kdeWrapper :: KDEPackage -> Package
```

What I say is that we can define an operation of combining two `KDEPackage`s:
```
mappend :: KDEPackage -> KDEPackage -> KDEPackage
mappend a b = { targets = a.targets ++ b.targets; paths = a.paths ++ b.paths; }
```

Notice that there's no symlinks involved -- we have just made a "package" consisting of merged sets of two packages.

Now, we'd like to allow `Package` result of `kdeWrapper` to be used as `KDEPackage` -- we can do that by means of `passthru`, exposing `targets` and `paths`. At this point we can merge `Package` using the same `mappend`, getting new `KDEPackage`s -- again, no symlinks, we merge initial sets of paths!

Finally let's combine `mappend` and `kdeWrapper` for convenience:

```
kdeWrapper :: [KDEPackage] -> Package
kdeWrapper pkgs =
  let targets = concatMap (p: p.targets) pkgs;
      paths = concatMap (p: p.paths) pkgs;
  in stdenv.mkDerivation {
    ...
    passthru = { inherit targets paths; };
  };
```

Notice that you can pass both manually-constructed `KDEPackage`s (like one defined in `dolphin.nix`) and results of previous `kdeWrapper`s (because of `passthru`) in that list -- all the targets and paths will be combined, and no symlinks are involved!

**Comment by ttuegel** [comment]
> Notice that you can pass both manually-constructed KDEPackages (like one defined in dolphin.nix) and results of previous kdeWrappers (because of passthru) in that list -- all the targets and paths will be combined, and no symlinks are involved!

I appreciate the time it took to write out this Haskell pseudo-code, but your example won't typecheck: the result of `kdeWrapper` is a `Package` not a `KDEPackage`. `KDEPackage` is not a Nix derivation, but merely an attribute set; it's not something you can install and it's not something you can put in the `paths` list. Now, you can create a `Package` from a `KDEPackage`, but that creates an extra build in the Nix store whose only purpose is to serve as an indirection. 

It's still a good idea to allow `kdeWrapper` to take a list of packages! It's just not a semigroup because it's not associative.

**Comment by abbradar** [comment]
Hm, I still feel that there is some misunderstanding. My idea is to _allow_ `Package` (Nix derivation) to pose as `KDEPackage` (attribute set with `targets` and `paths`) via `passthru`. Let's name something else, `InstallableKDEPackage` -- a `Package` that also contains `targets` and `paths` attributes taken _directly_ from converted `KDEPackage`. Now, we have a way to "lift" `KDEPackage` (a mere attribute set) to `InstallableKDEPackage` (Nix derivation with passed through `targets` and `paths` from the original `KDEPackage`):

```
lift :: KDEPackage -> InstallableKDEPackage
lift attrs = stdenv.mkDerivation {
  ...
  passthru = attrs;
}
```

AKA `kdeWrapper`. Notice again that this resulting `InstallableKDEPackage` would have the _same_ `targets` and `paths` copied from original `KDEPackage`!

Now we have a semigroup:

```
mappend :: InstallableKDEPackage -> InstallableKDEPackage -> InstallableKDEPackage
mappend pkg1 pkg2 = lift { paths = pkg1.paths ++ pkg2.paths; targets = pkg1.targets + pkg2.targets; }
```

I should have introduced `InstallableKDEPackage` in the first post but I was afraid it'd complicate the whole idea. Now I see that I shouldn't leave the types out when describing ~_^.

Now, Nix is not really static-typed so we can combine `lift` with `mconcat` and get the resulting `kdeWrapper` example of which I've posted above.

**Comment by abbradar** [comment]
I've implemented this idea in https://github.com/abbradar/nixpkgs/commit/dc17b9b4b01f8be2fa7ac62b675be4e4ddab1e29. This is not ideal (`/share` contains everything from `paths`, not just from `unwrapped`) but close to what I envisioned. You can compare:

`nix-build -A kde5.okular`

and

`nix-build -E 'with import <nixpkgs> {}; kde5.kdeWrapper (with kde5; [ okular konsole ])'`

Notice that in second case there won't be an extra symlink layer.

**Comment by FRidh** [comment]
@abbradar nice! 

I guess the next step is to apply this to the packages that are put in `environment.systemPackages` at https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/desktop-managers/kde5.nix
. I think it would be good to have an `kde5.extraPackages` option that adds those to the env before adding the env to `environment.systemPackages`.

**Comment by abbradar** [comment]
@ttuegel Your thoughts on this? If this seems a good idea overall I can make a PR.

**Comment by ttuegel** [comment]
@abbradar It sounds good to me. Please do that!

**Comment by abbradar** [comment]
Can we close this now?

**Comment by ttuegel** [comment]
Yes, everything should be fine now. Thanks!

**Comment by olejorgenb** [comment]
Starting okular 5 is still slow and stats a zillion files. 2-4 seconds with SSD disk.

Running strace and grepping for "stat(.*icon" i get **433078** stat calls. (`XDG_DATA_DIRS` have 9 entries)

https://woboq.com/blog/qicon-reads-gtk-icon-cache-in-qt57.html mentions something about an optimization for icon lookup using the gtk icon cache

>How do I make sure the cache is used?
>
> Every time you install an application that adds or removes icons from one of the themes, the cache needs to be updated. Qt looks at the modification time of the directories, if any directory is newer than the cache, the cache will not be used. You need to run gtk-update-icon-cache after installing new icons, but most distributions take care of this for you.

**Comment by bgamari** [comment]
I am still seeing quite long start-up time due to icon searching on NixOS 20.09:
```
$ nix run nixpkgs.kdeApplications.kate -c strace --summary --timestamps=unix,ms -o log kate
$ tail log
...
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- -------------------
 84.58    1.125760           4    251142    250137 access
  4.81    0.063955           4     14939     10959 statx
  2.17    0.028833           4      6838       747 stat
  2.00    0.026677           5      4526      2667 openat
  1.59    0.021119           3      5753         8 lstat
  1.52    0.020274          10      1879           close
  0.65    0.008661           7      1107       212 futex
...
```

**Comment by vs49688** [comment]
I'm also seeing this with multiple machines on `22.05pre365448.1063244793d`.

kate:
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- -------------------
 68.96    0.524305           2    212286    198879 statx
  9.22    0.070089           2     32874     30061 readlink
  6.62    0.050304           3     16336      1420 newfstatat
  4.01    0.030460           3      9786      4088 access
  2.47    0.018805           2      7003      4120 openat
  2.34    0.017766           5      3094         2 read
  1.92    0.014630        4876         3           wait4
  0.91    0.006915           5      1350       289 futex
  0.80    0.006065           2      2919           close
```

konsole:
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- -------------------
 36.86    0.032791           2     13091     12020 statx
 12.89    0.011472           2      4084      3359 openat
  9.37    0.008339        8339         1           wait4
  4.91    0.004372           3      1457      1163 access
  4.78    0.004254           1      2217      1372 newfstatat
  4.62    0.004108           9       449       128 futex
  4.51    0.004013           2      1891      1674 readlink
  4.12    0.003670           3      1034           mmap
  2.77    0.002465           1      1650           getpid
```

I suspect this is related to `XDG_DATA_DIRS` which is `6398` characters long. With konsole, clearing it reduces the amount of `statx` calls to `5656` (and starts faster). The konsole wrapper itself adds another `3345` characters, bringing the total to `9743`. 

This is almost certainly related to https://github.com/NixOS/nixpkgs/issues/126590
